Updates to contrib guidelines (#30)
cloudposse/.github: CONTRIBUTING.md

Unsure where to begin contributing to Cloud Posse? You can start by looking through issues with the following labels:

| Label                                                             | Usage                                                                    |
| :---------------------------------------------------------------- | :----------------------------------------------------------------------- |
| ![`help-wanted`](https://img.shields.io/badge/help_wanted-388bfd) | issues which should only require a few lines of code, and a test or two. |

> | ![`beginner`](https://img.shields.io/badge/beginner-388bfd)       | issues which should be a bit more involved than issues.                  |

Reviewer: As a side note: please add `?style=for-the-badge` to the end of shield image URLs, as in https://img.shields.io/badge/help_wanted-388bfd?style=for-the-badge

It is a new look we are adopting.

Please follow these steps to have your contribution considered by the maintainers:

1. Follow all instructions in [Pull Request template](.github/PULL_REQUEST_TEMPLATE.md)
2. Follow the [styleguides](#styleguides)
3. After you submit your pull request, verify that all status checks are passing
4. If you need to request a review on your PR, please do so in the #pr-reviews Slack channel

While the prerequisites above must be satisfied prior to having your pull request reviewed, the reviewer(s) may ask you to complete additional design work, tests, or other changes before your pull request can be ultimately accepted.

#### Versioning

The release automation relies on labels to increment the semantic versioning tag.

> As a reference, the following is a mapping between labels and semver:

Reviewer: ```suggestion
As a reference, the following is a mapping between labels and [semver](https://semver.org/):
```

> 1. `Patch`: A minor, backward compatible change. Increments patch version, e.g. 1.0.**1**

Reviewer 2: there is likely some automation that can help with the semver (sugesting) bit

Author: @hans-d that would be awesome. Have you seen any suitable actions?

Reviewer: We use [release-drafter](https://github.com/release-drafter/release-drafter) to automatically generate version numbers based on labels applied to the PR. All you have to do is add the appropriate labels. If there are no labels, "minor" is used.

We do not have automation to determine the labels. 

We also still have a lot of modules on v0, which has different rules. Only Cloud Posse engineers should move a module from v0 to v1, and usually we prefer that it not be a breaking change. In fact, it's not uncommon for v1 to be just a new label for the current release, prompted by the desire to release a v2 that has a breaking change.

My point of view is that the primary reason for adhering to [semver](https://semver.org/) is automation. If a change is made that for some reason should not be applied without a human review, then that is a Major change. According to semver, if there is a breaking change, that should be a Major change, and usually we follow that, but we might make a major release due to a change in policy (such as changing default behavior).

On the other side, it should always be safe to apply a patch version automatically without review. If a change fixes a bug in a way that might cause a service interruption or other problem for people relying on the buggy behavior, then that may be a minor or even major change, even though bugfixes can usually be released as patches. 

In the end it is a balance between wanting to warn people the release needs manual review, and not wearing people out with reviews of trivial changes.

Author: Thanks @Nuru
I have made some updates based on your comment, but you may want to add more in a future change.
Sometimes I find it difficult to determine the right label to use, so I think this section could benefit from some examples or at least better descriptions of each label.

2. `Minor`: New features that do not break anything, e.g. 1.**1**.1
3. `Major`: Breaking changes (or first stable release), e.g. **2**.0.0
4. `No-release`: Do not release a new version

Typically when a module is pre-release, e.g. 0.x.x, each new feature will be a minor release. Once the module hits 1.x, simpler features, which fit the definition of patch, will be labelled as such.

#### Stale Issues/PRs

It is proposed that stale issues/PRs are labelled `stale` after 30 days, and closed after 60 days if the `stale` label is present.

A Slack notification will be sent when an issue/PR is labelled `stale`.

To handle exceptions, additional labels have been proposed to control treatment of stale issues. The `stale` label will not be applied if any of the following labels are applied to a PR.

> The currently proposed labels are:

Reviewer 3: I'd suggest we add an "On Hold" label as well that would allow us to mark a PR as not ready to merge, but we don't want to go into the stale/auto-close cycle.

Reviewer 2: Even having these labels should be monitored. Perhaps only delay the first "stale" mention, but otherwise would say: just run the stale bit anyway. We should be able to remove that label and start a new period to get it marked stale again instead.

Author: @Gowiem thanks for the review.
I think @osterman 's idea was to use the presence of any of the other labels in the list as an indicator that it shouldn't be closed due to staleness. But I may have misunderstood.

@hans-d yeah that's definitely worth considering otherwise the other labels are just an alt route to things never being dealt with!

Reviewer: We have the `wip` and `do not merge` labels for PRs under active discussion and development. The `wip` (Work In Progress) label is for when the author does not consider the PR finalized, and has published it for comments and feedback on what is there but wants to add more before merging.

The `do not merge` label is for PRs in a state where releasing them would cause damage. It is along the lines of "changes requested", but can be used even when there are no proposed changes that would make the PR acceptable. 

I believe the standard label for preventing the "stale" label from being automatically applied is "frozen".

Author: @Nuru I'm thinking this section should be removed for now until a final decision is made.

All Terraform should adhere to our [Terraform Best Practices](https://docs.cloudposse.com/reference/best-practices/terraform-best-practices/).

## Terraform Module Contributions

### Updating Module Documentation

When updating variables, update the variable description if it could be made clearer, and especially if a new map/object attribute is added

To update the README and Terraform registry documentation, run the following commands:

```shell
make init
make github/init
make readme
```

> You may also need to install gomplate: https://docs.gomplate.ca/installing/

Reviewer 2: we should be able to auto-update those things, saves "oops, forgotten". And by automating it, we know it's the most recent version.

> You may also need to install gomplate: https://docs.gomplate.ca/installing/

Author: @Nuru [fixed](https://github.com/cloudposse/build-harness/pull/375) this yesterday. @Gowiem if you get a chance, please test the fix on your machine.

> You may also need to install gomplate: https://docs.gomplate.ca/installing/

Reviewer 3: I believe this is working now 👍 

Let's cut this line: 

```suggestion
```

### Updating the Module Examples

Each module has at least one example, located at `examples/complete/`. These are used as a guide for module consumers as well as for the Terratest inputs.

Where possible, update the example with representative values to include your changes in the test coverage.

### Deprecating a Variable

Sometimes variables have been defined with less than optimal types, or their usage changes over time due to provider/API changes.

To avoid introducing breaking changes, here is a suggested approach for deprecating a variable:

1. Add a file `variables-deprecated.tf` if it doesn't already exist
2. Move the deprecated variable into this file
3. Add your replacement variable
4. Configure the deprecated variable as a fallback of the replacement one, i.e. populate the value of the deprecated variable if the replacement variable is not supplied.

#### Examples:

- [terraform-aws-s3-bucket / variables-deprecated.tf](https://github.com/cloudposse/terraform-aws-s3-bucket/blob/main/variables-deprecated.tf)

### Optional Input Variables

Terraform does not easily allow conditional creation of resources based on input values.

> To avoid errors like `The "count" value depends on resource attributes that cannot be determined until apply time, so Terraform cannot predict how many instances will be created. To work around this, use the -target argument to first apply only the resources that the count depends on.` Cloud Posse has [standardised](https://github.com/cloudposse/terraform-aws-security-group/wiki/Naming-Conventions,-Deprecating-Inputs,-Optional-Inputs#optional-inputs) on using lists where inputs may be empty, and they determine the existence of another resource.

Reviewer 3: Good callout. 

When the same happens with `map` it's also a PITA and something that feels like we have a standard on. @aknysh described the go to approach here and I think he put it succinctly: https://sweetops.slack.com/archives/CB6GHNLG0/p1705692959914809?thread_ts=1705692678.037929&cid=CB6GHNLG0

Once you get this merged, I'll follow up with some docs on that.

Reviewer 2: can we apply standard/automated linting rules ?

Author: @Gowiem yes, it also happens with lists and maps. 
Re: that link, yes it would be great to have recommendations on the `The true and false result expressions must have consistent types` error too.

Author: @hans-d I think it might be difficult to automate these ones because the issues show up when the inputs depend on external state.

Reviewer 2: Linters should be able to identify such cases. but seeing that eg tflint does not have something for this (yet?)

Reviewer: We run linters automatically now and they post comments on the PR if they find issues.

Reviewer: It looks like I'm going to have to write a document explaining maps vs lists as inputs. I hope to get you a better link than that wiki chapter on optional inputs, too.

Author: During development of a change I often run the tests locally. I've added a change to capture your tips @Nuru and also a warning about orphaned resources.

This is because we are then able to use the length of the list as the conditional subject, rather than the value itself.

If you hit this case, define the variable using a list of the type you need, with the default `[]`. Add a validation condition to prohibit more than 1 element in the list.

#### Examples:

- [terraform-aws-security-group / variables.tf](https://github.com/cloudposse/terraform-aws-security-group/blob/main/variables.tf)
  ```hcl
    variable "target_security_group_id" {
            type        = list(string)

>             description = <<-EOT

Reviewer 2: Use of EOT/EOF for longer descriptions should also be a recommendation...

Author: @hans-d (thanks for the review!) it's included in https://docs.cloudposse.com/reference/best-practices/terraform-best-practices/#use-indented-heredoc-syntax

All Terraform should adhere to our [Terraform Best Practices](https://docs.cloudposse.com/reference/best-practices/terraform-best-practices/).

## Terraform Module Contributions

### Updating Module Documentation

When updating variables, update the variable description if it could be made clearer, and especially if a new map/object attribute is added

To update the README and Terraform registry documentation, run the following commands:

```shell

> make init
> make github/init
> make readme

Reviewer: We no longer use `make github/init`. 

```suggestion
make init
make readme
```

```

### Updating the Module Examples

Each module has at least one example, located at `examples/complete/`. These are used as a guide for module consumers as well as for the Terratest inputs.

Where possible, update the example with representative values to include your changes in the test coverage.

> 

Reviewer: There is also `go` test code using [terratest](https://github.com/gruntwork-io/terratest) under `test/src/` that may need to be updated, as the examples (always `examples/complete/`, usually other examples) are used for automated testing. 

Any changes to `versions.tf` in the module should also be made to `versions.tf` in `examples/complete/`

### Deprecating a Variable

Sometimes variables have been defined with less than optimal types, or their usage changes over time due to provider/API changes.

To avoid introducing breaking changes, here is a suggested approach for deprecating a variable:

1. Add a file `variables-deprecated.tf` if it doesn't already exist
2. Move the deprecated variable into this file

> 3. Add your replacement variable
> 4. Configure the deprecated variable as a fallback of the replacement one, i.e. populate the value of the deprecated variable if the replacement variable is not supplied.

Reviewer: ```suggestion
3. Add your replacement variable to `variables.tf`
4. Configure the deprecated variable as a fallback for the replacement one, i.e. use the value of the deprecated variable if it not the default value and the replacement variable is the default value. Otherwise, the replacement variable takes precedence.
```

  ```hcl
    variable "target_security_group_id" {
            type        = list(string)
            description = <<-EOT
            The ID of an existing Security Group to which Security Group rules will be assigned.
            The Security Group's name and description will not be changed.
            Not compatible with `inline_rules_enabled` or `revoke_rules_on_delete`.
            If not provided (the default), this module will create a security group.
            EOT
            default     = []
            validation {
            condition     = length(var.target_security_group_id) < 2
            error_message = "Only 1 security group can be targeted."
            }
        }
    ```

### Testing

Each module has Terratest tests in the `test/src` directory. Tests are written in Go.

#### Local Testing

Using [Leapp](https://docs.cloudposse.com/howto/geodesic/authenticate-with-leapp/), [aws-vault](https://docs.cloudposse.com/howto/geodesic/authenticate-with-aws-vault/) or or some other method, populate the following AWS credential environment variables for a suitable account:

- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`
- `AWS_SESSION_TOKEN`

```shell
cd test/src

> make docker/test

Reviewer 2: Good to know: wont work on arm / mac with apple sillicon

Author: @hans-d what behaviour do you see on this platform?

Reviewer 2: It directly fails, as no arm image is available. There seems to be a simple fix, but that code needs to be in each repo (unsure yet if there is a master repos where the test generics are pushed from)

Reviewer: We do not expect you to run the tests locally. They are run against your PR by Cloud Posse automation via GitHub actions, and in some cases need additional credentials or infrastructure, which Cloud Posse supples. Also, we are not rigorous about ensuring that resources are deleted if the test crashes (rare, but not impossible), falling back on a nightly scheduled cleanup. We would not want you to be paying for abandoned resources if the test should crash. 

If you do want to run the tests locally anyway, you can:

```
cd test/src
make test
```

This requires that you have the applicable version of `go` installed.

As @hans-d mentioned, our current test tooling Docker images are only available for Intel, and the Makefile does not automatically pull cross-platform images. However, Apple can emulate Intel, so if you want to run the test inside a Docker container on an Apple CPU, all you have to do is manually pull the test image first:

```
docker pull --platform linux/amd64 cloudposse/test-harness:latest
```

Reviewer 4: @Nuru I do not agree with this statement.

> We do not expect you to run the tests locally.

Reviewer 2: `--platform linux/amd64`  is something we can provide in the makefile (eg via env var). 
Now it needs to be done on a repo by repo base, but would rather have this a generic change.

Reviewer 2: > > We do not expect you to run the tests locally.
> It might have broke at some point, but it was always designed to run locally. When I was more active in development, used it all the time.

I prefer to be able to run all tests/checks/updates locally without having to rely on a pipeline, esp when someone needs to initiate that (as in these repos)

Reviewer 2: Good to include that running this locally is more a strong suggestion vs mandatory - as it will improve the review cycle - and certainly helps with the more tricky changes

Author: Good point @hans-d - I've added a note about this

To avoid introducing breaking changes, here is a suggested approach for deprecating a variable:

1. Add a file `variables-deprecated.tf` if it doesn't already exist

> 3. Add your replacement variable to `variables.tf`

Author: I used VS Code to review the PR and this part seemed to have gotten mangled. Should be OK now.
